core/src/main/java/org/geysermc/geyser/extension/command/GeyserExtensionCommand.java

```diff
@@ -25,19 +25,210 @@
+import org.geysermc.geyser.api.command.CommandSource;
+import org.geysermc.geyser.api.connection.GeyserConnection;
 import org.geysermc.geyser.api.extension.Extension;
+import org.geysermc.geyser.api.util.TriState;
 import org.geysermc.geyser.command.GeyserCommand;
+import org.geysermc.geyser.command.GeyserCommandSource;
+import org.geysermc.geyser.session.GeyserSession;
+import org.incendo.cloud.CommandManager;
+import org.incendo.cloud.context.CommandContext;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import static org.incendo.cloud.parser.standard.StringParser.greedyStringParser;
 
 public abstract class GeyserExtensionCommand extends GeyserCommand {
+
     private final Extension extension;
+    private final String rootCommand;
 
-    public GeyserExtensionCommand(Extension extension, String name, String description, String permission) {
-        super(name, description, permission);
+    public GeyserExtensionCommand(@NonNull Extension extension, @NonNull String name, @NonNull String description,
+                                  @NonNull String permission, @Nullable TriState permissionDefault,
+                                  boolean playerOnly, boolean bedrockOnly) {
 
+        super(name, description, permission, permissionDefault, playerOnly, bedrockOnly);
         this.extension = extension;
+        this.rootCommand = Objects.requireNonNull(extension.rootCommand());
+
+        if (this.rootCommand.isBlank()) {
+            throw new IllegalStateException("rootCommand of extension " + extension.name() + " may not be blank");
+        }
     }
 
-    public Extension extension() {
+    public final Extension extension() {
         return this.extension;
     }
+
+    @Override
+    public final String rootCommand() {
+        return this.rootCommand;
+    }
+
+    public static class Builder<T extends CommandSource> implements Command.Builder<T> {
+        @NonNull private final Extension extension;
+        @Nullable private Class<? extends T> sourceType;
+        @Nullable private String name;
+        @NonNull private String description = "";
+        @NonNull private String permission = "";
+        @Nullable private TriState permissionDefault;
+        @Nullable private List<String> aliases;
+        private boolean suggestedOpOnly = false; // deprecated for removal
+        private boolean playerOnly = false;
+        private boolean bedrockOnly = false;
+        @Nullable private CommandExecutor<T> executor;
+
+        public Builder(Extension extension) {
+            this.extension = Objects.requireNonNull(extension);
+        }
+
+        @Override
+        public Command.Builder<T> source(@NonNull Class<? extends T> sourceType) {
+            this.sourceType = sourceType;
+            return this;
+        }
+
+        @Override
+        public Builder<T> name(@NonNull String name) {
+            this.name = name;
+            return this;
+        }
```

Comment:
if this is nonnull, shouldn't it also check for a null/blank name being passed here instead of when building the command?